nifi-extension-bundles/nifi-kafka-bundle/nifi-kafka-processors/src/main/java/org/apache/nifi/kafka/processors/producer/convert/RecordWrapperStreamKafkaRecordConverter.java

```diff
@@ -107,7 +107,10 @@ public KafkaRecord next() {
                     final RecordFieldConverter converter = new RecordFieldConverter(record, flowFile, logger);
                     final byte[] key = converter.toBytes(WrapperRecord.KEY, keyWriterFactory);
                     final byte[] value = converter.toBytes(WrapperRecord.VALUE, writerFactory);
-                    ProducerUtils.checkMessageSize(maxMessageSize, value.length);
+
+                    if (value != null) {
```

Comment:
For the RecordWrapper strategy, there needs to be some logic at this point to do the evaluation of value and the FlowFile attribute "kafka.tombstone".  One option is to implement a utility method to do things inline at line 109, something like:

```
private byte[] toValue(final byte[] bytes, final FlowFile flowFile) {
    final String tombstone = flowFile.getAttribute(KafkaFlowFileAttribute.KAFKA_TOMBSTONE);
    final boolean isTombstone = (Boolean.TRUE.toString().equals(tombstone) && (flowFile.getSize() == 0));
    return isTombstone ? null : bytes;
}
```

This is surfaced by adding the zero byte value test case described in this review.

Reply:
Are you sure to depend a tombstone produce on both value=null and the flowfile attribute? I know, on the one hand, the attribute is a part of the "contract" of the nifi-kafka-interface, but on the other hand as a user, I would not assume to also set the flowfile attribute after adjusting the content to wrapper structure (e.g. jolt-transfom or others). Additionally, we can keep the code cleaner ;-)
I would follow your intention in this point. If you want to, I'll add the dependency!

Reply:
Yes, there is some ambiguity here.  

Given the other code touch, it seems like it makes sense to let the "wrapper record" value pass through, which would be implicitly interpreted by Kafka as a tombstone.  The confusion is that for the FlowFile strategy, it is not possible to set the value to null, so we need to flex based on the FlowFile attribute.  But not for this strategy.

I think this change is good...